core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -70,18 +70,206 @@ class SYCLInternal {
 
   std::unique_ptr<sycl::queue> m_queue;
 
+  // USMObjectMem is a reusable buffer for a single object
+  // in USM memory
+  template <sycl::usm::alloc Kind>
+  class USMObjectMem {
+   public:
+    class Deleter {
+     public:
+      Deleter() = default;
+      explicit Deleter(USMObjectMem* mem) : m_mem(mem) {}
+
+      template <typename T>
+      void operator()(T* p) const noexcept {
+        assert(m_mem);
+        assert(sizeof(T) == m_mem->m_size);
+
+        if constexpr (sycl::usm::alloc::device == kind)
+          // Only skipping the detor on trivially copyable types
+          static_assert(std::is_trivially_copyable_v<T>);
+        else
+          p->~T();
+
+        m_mem->m_size = 0;
+      }
+
+     private:
+      USMObjectMem* m_mem = nullptr;
+    };
+
+    static constexpr sycl::usm::alloc kind = Kind;
+
+    void reset() {
+      assert(!m_size);
+
+      if (m_data) {
+        sycl::free(m_data, *m_q);
+        m_capacity = 0;
+        m_data     = nullptr;
+      }
+      m_q.reset();
+    }
+
+    void reset(sycl::queue q) {
+      reset();
+      m_q.emplace(std::move(q));
+    }
+
+    USMObjectMem() = default;
+    explicit USMObjectMem(sycl::queue q) noexcept : m_q(std::move(q)) {}
+
+    USMObjectMem(USMObjectMem const&) = delete;
+    USMObjectMem(USMObjectMem&&)      = delete;
+    USMObjectMem& operator=(USMObjectMem&&) = delete;
+    USMObjectMem& operator=(USMObjectMem const&) = delete;
+
+    ~USMObjectMem() {
+      assert(!m_size);
+
+      if (m_data) sycl::free(m_data, *m_q);
+    }
+
+    void* data() noexcept { return m_data; }
+    const void* data() const noexcept { return m_data; }
+
+    size_t size() const noexcept { return m_size; }
+    size_t capacity() const noexcept { return m_capacity; }
+
+    // reserve() allocates space for at least n bytes
+    // returns the new capacity
+    size_t reserve(size_t n) {
+      assert(!m_size);
+      assert(m_q);
+
+      if (m_capacity < n) {
+        // First free what we have (in case malloc can reuse it)
+        sycl::free(m_data, *m_q);
+
+        m_data = sycl::malloc(n, *m_q, kind);
+        if (!m_data) {
+          m_capacity = 0;
+          throw std::bad_alloc();
```

Comment:
Possibly use a Kokkos function like `Kokkos::Impl::throw_runtime_exception()`?

Reply:
yeah

Reply:
More importantly, I think we should go through our memory spaces explicitly so that this memory gets tracked by our memory tracking. Or is there a good reason not to do that? I would like internal memory to show up as much as possible in the tools, so folks don't have to ask us where all the extra memory is coming from their apps need. In particular it would allow us to be reasonable confident about what is caused by us vs the runtime of the underlying programming model.

Reply:
I agree with that, although I'd rather do the change to use memory spaces under a separate P/R, as those changes have to happen to parallel scan as well.

Reply:
Then mark it with FIXME_SYCL so we don't forget about it.